### PR TITLE
CosmosOperationCanceledException: Adds serializable functionality

### LINKED ISSUE #3402: Make CosmosOperationCanceledException Serializable
My team just upgraded from 3.20.1 to 3.29.0 and we are now running into problems because `CosmosOperationCanceledException` no longer works  with the [Microsoft Orleans binary serializer](https://docs.microsoft.com/en-us/dotnet/orleans/host/configuration-guide/serialization#exception-serialization).

Azure-sdk-for-net ran into this [same issue](https://github.com/Azure/azure-sdk-for-net/issues/10316) a few years ago, and solved it by implementing `Serializable` on their `Exception` classes.

It's a best practice to always make `Exception` classes `Serializable`. Without this, my team has to work around the issue by catching all CosmosExceptions and wrapping them in another Exception that implements Serializable.

### PR DESCRIPTION
**Pull request by NaluTripician**
# Pull Request Template

## Description

Code changes to make the `CosmosOperationCanceledException` Exception serializable. It should be noted that `CosmosDiagnostics` are not serializable and are left out of sterilization and upon deserialization the original `CosmosDiagnotics` are replaced with a `CosmosTraceDiagnostics` with a `NoOpTrace`. `lazyMessage` and `toStringMessage` for deserialized `CosmosOperationCanceledException`s are also updated to inform users that `CosmosDiagnostics` cannot be deserialized. Additionally `CosmosOperationCanceledTests.cs` was also created to test that the relevant information can be serialized and recovered. 

## Type of change

Please delete options that are not relevant.

- [] New feature (non-breaking change which adds functionality)
- [] This change requires a documentation update

## Closing issues

To automatically close an issue: closes #3402

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/Resource/CosmosExceptions/CosmosOperationCanceledException.cs [hunk 1]
```diff
@@ -112,6 +134,15 @@ private Lazy<string> CreateToStringMessage()
             return new Lazy<string>(() => $"{this.originalException}{Environment.NewLine}Cancellation Token has expired: {this.tokenCancellationRequested}. Learn more at: https://aka.ms/cosmosdb-tsg-request-timeout{Environment.NewLine}CosmosDiagnostics: {this.Diagnostics}");
         }
 
+        private Lazy<string> CreateLazyMessageDeserialized()
+        {
+            return new Lazy<string>(() => $"{this.originalException.Message}{Environment.NewLine}Cancellation Token has expired: {this.tokenCancellationRequested}. Note that CosmosDiagnostics cannot be serialized. Learn more at: https://aka.ms/cosmosdb-tsg-request-timeout{Environment.NewLine}CosmosDiagnostics: {this.Diagnostics}");
+        }
+        private Lazy<string> CreateToStringMessagDeserialized()
+        {
+            return new Lazy<string>(() => $"{this.originalException}{Environment.NewLine}Cancellation Token has expired: {this.tokenCancellationRequested}. Note that CosmosDiagnostics cannot be serialized. Learn more at: https://aka.ms/cosmosdb-tsg-request-timeout{Environment.NewLine}CosmosDiagnostics: {this.Diagnostics}");
```
Comment (sourabh1007) [inline Microsoft.Azure.Cosmos/src/Resource/CosmosExceptions/CosmosOperationCanceledException.cs:143]: Why are you referring to timeout documentation here?

Comment (NaluTripician) [reply Microsoft.Azure.Cosmos/src/Resource/CosmosExceptions/CosmosOperationCanceledException.cs:143]: Here I was copying the message from the original `CreateToString()` Method. The only thing I changed with this new method was adding “Note that CosmosDiagnostics cannot be serialized” to the returned Lazy<string>

File: Microsoft.Azure.Cosmos/src/Resource/CosmosExceptions/CosmosOperationCanceledException.cs [hunk 2]
```diff
@@ -123,5 +154,17 @@ internal static void RecordOtelAttributes(CosmosOperationCanceledException excep
...
+        /// <summary>
+        /// Sets the System.Runtime.Serialization.SerializationInfo with information about the exception.
+        /// </summary>
+        /// <param name="info">The SerializationInfo object that holds serialized object data for the exception being thrown.</param>
+        /// <param name="context">The StreamingContext that contains contextual information about the source or destination.</param>
+        public override void GetObjectData(SerializationInfo info, StreamingContext context)
+        {
+            base.GetObjectData(info, context);
+            info.AddValue("originalException", this.originalException);
+            info.AddValue("tokenCancellationRequested", this.tokenCancellationRequested);
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Resource/CosmosExceptions/CosmosOperationCanceledException.cs:167]: Why not serialize here the `lazyMessage` and `toStringMessage`:

```
info.AddValue("lazyMessage", this.lazyMessage.Value);
info.AddValue("toStringMessage", this.toStringMessage.Value);
```

And then on the other method:

```
this.lazyMessage = new Lazy<string>(() => (string)info.GetValue("lazyMessage", typeof(string)));
this.toStringMessage= new Lazy<string>(() => (string)info.GetValue("toStringMessage", typeof(string)));
```

That way you don't need the new methods.

Comment (NaluTripician) [reply Microsoft.Azure.Cosmos/src/Resource/CosmosExceptions/CosmosOperationCanceledException.cs:167]: This is definitely possible; do we still want to have the message stating that `CosmosDiagnostics` cannot be serialized? If so, do we only want that part of the message when it is deserialized or all of the time? I can just modify the message after it is deserialized if it is the former.

Comment (ealsur) [reply Microsoft.Azure.Cosmos/src/Resource/CosmosExceptions/CosmosOperationCanceledException.cs:159]: Not needed in the actual message, just leave a comment on the code, when the object is deserialized:

```
// Diagnostics cannot be serialized
this.Diagnostics = new CosmosTraceDiagnostics(NoOpTrace.Singleton);
```

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/CosmosOperationCanceledTests.cs [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+                new CosmosTraceDiagnostics(NoOpTrace.Singleton));
+
+            //serialize exception
+            string serialized = JsonConvert.SerializeObject(originalException);
+
+            CosmosOperationCanceledException deserializedExceptoin = 
+                JsonConvert.DeserializeObject<CosmosOperationCanceledException>(serialized);
+
+            //Asserts
+            Assert.AreEqual(originalException.ToString(), deserializedExceptoin.ToString());
```
[the remark this answers is not in the document]
Comment (ealsur) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/CosmosOperationCanceledTests.cs:38]: The Assert is not yet pushed, could it be pending?

[the remark this answers is not in the document]
Comment (NaluTripician) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/CosmosOperationCanceledTests.cs:38]: It has just been pushed.

File: Microsoft.Azure.Cosmos/src/Resource/CosmosExceptions/CosmosOperationCanceledException.cs [hunk 4]
```diff
@@ -6,18 +6,25 @@ namespace Microsoft.Azure.Cosmos
 {
     using System;
     using System.Collections;
+    using System.Diagnostics;
+    using System.Drawing.Printing;
```
Comment (ealsur) [inline Microsoft.Azure.Cosmos/src/Resource/CosmosExceptions/CosmosOperationCanceledException.cs:10]: Do you need all these `usings`? Could you clean them up to only maintain those you are referencing?

Comment (NaluTripician) [reply Microsoft.Azure.Cosmos/src/Resource/CosmosExceptions/CosmosOperationCanceledException.cs:10]: Cleaned them up, some were there before I started working on this file.

**Review by ealsur — APPROVED** [review]

**Comment by ealsur** [comment]
@NaluTripician See the test failures:

```
 Failed ContractChanges [179 ms]
  Error Message:
   Assert.Fail failed. Public API has changed. If this is expected, then run (EnlistmentRoot)\UpdateContracts.ps1 . To see the differences run the update script and use git diff.
  Stack Trace:

``` 

Because you are modifying a public type, this generates a change in the API contract, please run the Powershell the test indicates that should be in the root of your clone.

**Review by ealsur — APPROVED** [review]

**Pull request merged**